### PR TITLE
openapi: Improve operationId/request/response naming strategy

### PR DESCRIPTION
**Pull request by averche**
## Background

This PR changes the way we construct the `operationId` as well as the request & response names  in the generated `openapi.json` document. `operationId`s are translated directly into method names in the two new OpenAPI-based client libraries:

- [hashicorp/vault-client-go](https://github.com/hashicorp/vault-client-go/)
- [hashicorp/vault-client-dotnet](https://github.com/hashicorp/vault-client-dotnet/)

## Old naming strategy

Today the approach is to naively construct `operationId` values from the API paths. Unfortunately, the resultant methods names are not very human-friendly. Here are a couple examples:

```go
// notice roleRoleRole stutter and the redundant 'Auth' in the method name
client.Auth.PostAuthApproleRoleRoleName(...)

// this is the function generated for fetching KVv2 secrets (probably the most popular Vault use-case)
// but it sounds like it's getting a secret path instead!
client.Secrets.GetSecretDataPath(...)
```

Additionally, the request & response names had a different naming logic, which resulted in some naming collisions (https://github.com/hashicorp/vault/issues/18578).

## New naming strategy

The new naming strategy is to take advantage of the existing optional `DisplayAttributes` at Path and PathOperation level. I'm introducing a couple new `DisplayAttributes` fields:

https://github.com/hashicorp/vault/blob/72c4acd2d43353cdec4b04330a778b764ef535b0/sdk/framework/path.go#L227-L243

These fields will be used to help construct the operation ID / request / response names if defined. The fields will now be generated as `hyphenated-lowercase-strings` for better compatibility with various language generators.

For examples of how the new naming strategy works, please take a look at the new test:

https://github.com/hashicorp/vault/blob/72c4acd2d43353cdec4b04330a778b764ef535b0/sdk/framework/openapi_test.go#L571-L744

## Related work

This PR is a new approach based on some valid concerns brought forward by @maxb in a related #18321.

The changes to define display attributes for specific plugins will be done in upcoming related PRs:

- https://github.com/hashicorp/vault/pull/19363
- https://github.com/hashicorp/vault/pull/19366
- https://github.com/hashicorp/vault/pull/19383
- https://github.com/hashicorp/vault/pull/19384 
- https://github.com/hashicorp/vault/pull/19389
- https://github.com/hashicorp/vault/pull/19391
- https://github.com/hashicorp/vault/pull/19392
- https://github.com/hashicorp/vault/pull/19399
- https://github.com/hashicorp/vault/pull/19411 
- https://github.com/hashicorp/vault/pull/19413
- https://github.com/hashicorp/vault/pull/19414
- https://github.com/hashicorp/vault/pull/19415
- https://github.com/hashicorp/vault/pull/19422

### CODE REVIEW

**Review by AnPucel — COMMENTED** [review]
I looked at the test, but could we get examples in the description of what the new OperationId's will map to that are actual paths right now? E.g. `PostAuthApproleRoleRoleName` will now look like... ?

**Comment by averche** [comment]
After this PR is merged, the names are not going to be in an ideal form. For instance, `PostAuthApproleRoleRoleName` will turn into `ApproleWriteRoleRoleName`.

The follow-up PR(s) will populate the `DisplayAttributes` for AppRole and other plugins, so the final state will be close to what we have in the client libraries (`AppRoleWriteRole`).

File: sdk/framework/openapi.go
Comment (maxb) [inline sdk/framework/openapi.go:961]: Typo: 
```suggestion
// will soon be removed.
```

but also, can this really be removed any time soon?

Should it not really remain until all plugins built using older `sdk` versions are wholly unsupported, for which, AFAIK, there is no written policy - so not soon, at least?

Also, this function currently has the role of fixing up duplicate operation IDs by appending a numeric suffix, and that is now being bypassed when pre-calculated OperationIDs are used. Ideally, that functionality would be retained - because, even though it generally should not happen, humans occasionally make mistakes, and it's a lot easier to find them and correct them if the duplicate operation IDs are being visibly given numeric suffixes, rather than silently overwriting each other's request/response schemas.

File: sdk/framework/openapi.go
Comment (maxb) [inline sdk/framework/openapi.go:580]: Is this actually needed? I see you have support for defining the suffix at both path and operation levels, so the suffix could be written in singular form on the path DisplayAttrs, and then overridden to plural form in the ListOperation DisplayAttrs? That would (I think?) allow the same end goal without needing to hardcode particular parameter names here.

**Comment by maxb** [comment]
> For instance, `PostAuthApproleRoleRoleName` will turn into `ApproleWriteRoleRoleName`.

There are quite a few cases where there is an auth method and a secrets engine of the same name (e.g AWS, Azure, Kubernetes, LDAP). For this reason, I think it might be better to retain the `Auth` in the default prefix of auth methods - otherwise operations of these auth methods and secrets engines will mingle, and in the worse case, possibly even clash.

File: sdk/framework/openapi.go
Comment (averche) [reply sdk/framework/openapi.go:972]: Fair point, I suppose it will just have to stay here for a while, but will be marked deprecated, I'll update the comment.

File: sdk/framework/openapi.go
Comment (averche) [reply sdk/framework/openapi.go]: This was my first attempt at addressing the one to many ambiguity introduced by the `expandPattern` (https://github.com/hashicorp/vault/pull/18321#discussion_r1063235356).

I have since come up with a (hopefully) better solution in d3e16aaf967520a92cf20dd3090321603964683a. Let me know what you think :)

**Comment by averche** [comment]
> There are quite a few cases where there is an auth method and a secrets engine of the same name (e.g AWS, Azure, Kubernetes, LDAP). For this reason, I think it might be better to retain the Auth in the default prefix of auth methods - otherwise operations of these auth methods and secrets engines will mingle, and in the worse case, possibly even clash.

My plan is to address such ambiguities in related PR's by adding "Auth" into the `OperationSuffix` where necessary. Example: #19366. I will merge all PR's together once they are all reviewed & approved.

File: sdk/framework/openapi.go
Comment (averche) [reply sdk/framework/openapi.go]: > Also, this function currently has the role of fixing up duplicate operation IDs by appending a numeric suffix, and that is now being bypassed when pre-calculated OperationIDs are used. Ideally, that functionality would be retained - because, even though it generally should not happen, humans occasionally make mistakes, and it's a lot easier to find them and correct them if the duplicate operation IDs are being visibly given numeric suffixes, rather than silently overwriting each other's request/response schemas.

Currently, my plan is to address all such potential ambiguities in related PR's that adjust `OperationPrefix  / OperationSuffix / Action` for various built-in plugins. OpenAPI does not permit duplicate `operationId` values, so in case of duplicates (at least within the built-in plugins), the spec generated using [gen_openapi.sh](https://github.com/hashicorp/vault/blob/main/scripts/gen_openapi.sh) will be invalid and will be flagged by one of the CI jobs that we are planning to add in the future.

File: sdk/framework/openapi.go
Comment (dhuckins) [inline sdk/framework/openapi.go:602]: +1

File: sdk/framework/path.go
Comment (dhuckins) [inline sdk/framework/path.go:234]: what's the default if empty? we go off of http verb? (eg: POST requests are Write, etc)

**Review by dhuckins — APPROVED** [review]

File: sdk/framework/path.go
Comment (averche) [reply sdk/framework/path.go:234]: The default will be the Operation enum's string representation (e.g. read, delete, list, etc.). The only exception to that is "update" is replaced with "write" since it's more appropriate in most cases. The logic is at the end of `constructOperationID`

**Pull request merged**
